json_span.go

```diff
@@ -1,39 +1,227 @@
+}
+
+// HTTPSpanData represents the `data` section of an HTTP span sent within an OT span document
+type HTTPSpanData struct {
+	SpanData
+	Tags HTTPSpanTags `json:"http"`
+}
+
+// NewHTTPSpanData initializes a new HTTP span data from tracer span
+func NewHTTPSpanData(span *spanS) HTTPSpanData {
+	data := HTTPSpanData{
+		SpanData: NewSpanData(span, RegisteredSpanType(span.Operation)),
+		Tags:     NewHTTPSpanTags(span),
+	}
+
+	return data
+}
+
+// HTTPSpanTags contains fields within the `data.http` section of an OT span document
+type HTTPSpanTags struct {
+	URL      string `json:"url,omitempty"`
+	Status   int    `json:"status,omitempty"`
+	Method   string `json:"method,omitempty"`
+	Path     string `json:"path,omitempty"`
+	Host     string `json:"host,omitempty"`
+	Protocol string `json:"protocol,omitempty"`
+	Error    string `json:"error,omitempty"`
+}
+
+// NewHTTPSpanTags extracts HTTP-specific span tags from a tracer span
+func NewHTTPSpanTags(span *spanS) HTTPSpanTags {
+	var tags HTTPSpanTags
+	for k, v := range span.Tags {
+		switch k {
+		case "http.url", string(ext.HTTPUrl):
+			readStringTag(&tags.URL, v)
+		case "http.status", "http.status_code":
+			readIntTag(&tags.Status, v)
+		case "http.method", string(ext.HTTPMethod):
+			readStringTag(&tags.Method, v)
+		case "http.path":
+			readStringTag(&tags.Path, v)
+		case "http.host":
+			readStringTag(&tags.Host, v)
+		case "http.protocol":
+			readStringTag(&tags.Protocol, v)
+		case "http.error":
+			readStringTag(&tags.Error, v)
+		}
+	}
+
+	return tags
+}
+
+// RPCSpanData represents the `data` section of an RPC span sent within an OT span document
+type RPCSpanData struct {
+	SpanData
+	Tags RPCSpanTags `json:"rpc"`
+}
+
+// NewRPCSpanData initializes a new RPC span data from tracer span
+func NewRPCSpanData(span *spanS) RPCSpanData {
+	data := RPCSpanData{
+		SpanData: NewSpanData(span, RegisteredSpanType(span.Operation)),
+		Tags:     NewRPCSpanTags(span),
+	}
+
+	return data
+}
+
+// RPCSpanTags contains fields within the `data.rpc` section of an OT span document
+type RPCSpanTags struct {
+	Host     string `json:"host,omitempty"`
+	Port     string `json:"port,omitempty"`
+	Call     string `json:"call,omitempty"`
+	CallType string `json:"call_type,omitempty"`
+	Flavor   string `json:"flavor,omitempty"`
+	Error    string `json:"error,omitempty"`
+}
+
+// NewRPCSpanTags extracts RPC-specific span tags from a tracer span
+func NewRPCSpanTags(span *spanS) RPCSpanTags {
+	var tags RPCSpanTags
+	for k, v := range span.Tags {
+		switch k {
+		case "rpc.host":
+			readStringTag(&tags.Host, v)
+		case "rpc.port":
+			readStringTag(&tags.Port, v)
+		case "rpc.call":
+			readStringTag(&tags.Call, v)
+		case "rpc.call_type":
+			readStringTag(&tags.CallType, v)
+		case "rpc.flavor":
+			readStringTag(&tags.Flavor, v)
+		case "rpc.error":
+			readStringTag(&tags.Error, v)
+		}
+	}
+
+	return tags
+}
+
+func readStringTag(dst *string, tag interface{}) {
+	if v, ok := tag.(string); ok {
```

Comment:
Can it happen that the tag may not be string (or the expected type)?

Reply:
The `dst` won't be populated. The problem is that the OT span tags are not strictly typed, so we have to do our best. However it might make sense to handle `[]byte` here as well and various `int` types in the `readIntTag`. What do you think?

Reply:
Ok as long as we don't crash/error out.  Missing tags can be fixed if found later...